api/key/key.go

```diff
@@ -96,30 +97,47 @@ func Uint(k string, v uint) core.KeyValue {
 // Infer creates a new key-value pair instance with a passed name and
 // automatic type inference. This is slower, and not type-safe.
 func Infer(k string, value interface{}) core.KeyValue {
-	switch v := value.(type) {
-	case bool:
-		return Bool(k, v)
-	case int:
-		return Int(k, v)
-	case uint:
-		return Uint(k, v)
-	case int32:
-		return Int32(k, v)
-	case int64:
-		return Int64(k, v)
-	case uint32:
-		return Uint32(k, v)
-	case uint64:
-		return Uint64(k, v)
-	case float32:
-		return Float32(k, v)
-	case float64:
-		return Float64(k, v)
-	case string:
-		return String(k, v)
-	case fmt.Stringer:
-		return Stringer(k, v)
-	default:
-		return String(k, fmt.Sprint(v))
+	if value == nil {
+		return String(k, "<nil>")
 	}
+
+	if stringer, ok := value.(fmt.Stringer); ok {
+		return String(k, stringer.String())
+	}
+
+	rv := reflect.ValueOf(value)
+
+	switch rv.Kind() {
+	case reflect.Bool:
+		return Bool(k, rv.Interface().(bool))
+	case reflect.Int:
+		return Int(k, rv.Interface().(int))
```

Comment:
You are supposed to use `int(rv.Int())` here and everywhere else. This will make this version slightly more performant.

Reply:
Thanks! Done.

Reply:
(I added the new benchmark you gave above.)